GPU/Common/FramebufferManagerCommon.h

```diff
@@ -62,48 +63,76 @@ class VulkanFBO;
 // when such a situation is detected. In order to reliably detect this, we separately track depth buffers,
 // and they know which color buffer they were used with last.
 struct VirtualFramebuffer {
+	Draw::Framebuffer *fbo;
+
 	u32 fb_address;
 	u32 z_address;  // If 0, it's a "RAM" framebuffer.
-	int fb_stride;
-	int z_stride;
-
-	GEBufferFormat format;  // virtual, in reality they are all RGBA8888 for better quality but we can reinterpret that as necessary
+	u16 fb_stride;
+	u16 z_stride;
 
 	// width/height: The detected size of the current framebuffer, in original PSP pixels.
 	u16 width;
 	u16 height;
 
 	// bufferWidth/bufferHeight: The pre-scaling size of the buffer itself. May only be bigger than or equal to width/height.
-	// Actual physical buffer is this size times the render resolution multiplier.
+	// In original PSP pixels - actual framebuffer is this size times the render resolution multiplier.
 	// The buffer may be used to render a width or height from 0 to these values without being recreated.
 	u16 bufferWidth;
 	u16 bufferHeight;
 
 	// renderWidth/renderHeight: The scaled size we render at. May be scaled to render at higher resolutions.
-	// The physical buffer may be larger than renderWidth/renderHeight.
+	// These are simply bufferWidth/Height * renderScaleFactor and are thus redundant.
 	u16 renderWidth;
 	u16 renderHeight;
 
-	float renderScaleFactor;
+	// Attempt to keep track of a bounding rectangle of what's been actually drawn. However, right now these are not
+	// used in a very detailed way (they're only ever 0 or equal to width/height)
+	u16 drawnWidth;
+	u16 drawnHeight;
+
+	// The dimensions at which we are confident that we can read back this buffer without stomping on irrelevant memory.
+	u16 safeWidth;
+	u16 safeHeight;
+
+	// The scale factor at which we are rendering (to achieve higher resolution).
+	u8 renderScaleFactor;
+
+	// The original PSP format of the framebuffer.
+	// In reality they are all RGBA8888 for better quality but this is what the PSP thinks it is. This is necessary
+	// when we need to interpret the bits directly (depal or buffer aliasing).
+	GEBufferFormat format;
+
+	// The configured buffer format at the time of the latest/current draw. This will change first, then
+	// if different we'll "reinterpret" the framebuffer to match 'format' as needed.
+	GEBufferFormat drawnFormat;
 
 	u16 usageFlags;
 
+	// These are used to track state to try to avoid buffer size shifting back and forth.
+	// Though that shouldn't really happen, should it, since we always grow, don't shrink?
```

Comment:
We do resize down, if the size has shrunk for a while and the framebuffer is also larger than the stride.  At this point, the "safe" size is probably a lie, and I think we had various issues with readbacks, so this resizes down to avoid them.

An example would be a game that always uses the address 0x00154000 for temp buffers, and uses it for a full-screen effect for 3 frames, then goes back to using it for character shadows or something much smaller.

-[Unknown]

Reply:
I see, that makes sense. I'll revise the comment.